fix!: `NativeList<T>` causes crashes when used in RPCs [NCCBUG-106]

ShadauxCat opened the pull request:
There were two issues here:
- Any generic type that was passed to an RPC would result in a native crash because the IL code generated was invalid
- Even with valid IL code, passing NativeList to an RPC would result in a crash because the value being sent over the wire would just be the pointer address, which would not point to valid memory on the receiving system

To address this, two things have been done:
- The ILPP code has been fixed so it generates correct IL code with generics
- RPCs (and FastBufferWriter/Reader) are now limited in what they'll accept. Specifically, they'll accept:
  - Primitive types (`int`, `bool`, `byte`, etc)
  - `unmanaged` enums
  - Structs or classes that implement `INetworkSerializable`
  - `unmanaged` structs that are tagged as `ISerializeByMemcpy` (an empty interface that the developer can add to their struct to indicate that it's compatible with memcpy serialization and can be sent over the wire)
  - Unity built-in struct types (`Vector3`, `Color`, etc)
  - Any type for which extension methods for `FastBufferWriter` and `FastBufferReader` have been created to enable them to be serialized

For types that aren't natively supported but aren't in the developer's control to be able to add `ISerializeByMemcpy` to it, there is `ForceSerializeByMemcpy`, a wrapper that adds the tag while supporting implicit conversion to and from its wrapped type. It has the limitation that, in order to be supported by NativeList, the type it wraps must be both `unmanaged` and `IEquatable<T>`.

Currently, `FixedString` is not in the list of natively-supported types (due to the number of varying sizes it can have) and must be wrapped in `ForceSerializeByMemcpy` to be used.

## Changelog

- Fixed: Passing generic types to RPCs no longer causes a native crash
- Changed: `unmanaged` structs are no longer universally accepted as RPC parameters because some structs (i.e., structs with pointers in them, such as `NativeList<T>`) can't be supported by the universal struct serializer. Structs that are intended to be serialized across the network must add `ISerializeByMemcpy` to the interface list (i.e., `struct Foo : ISerializeByMemcpy`). This interface is empty and just serves to mark the struct as compatible with universal serialization. For external structs you can't edit, you can pass them to RPCs by wrapping them in `ForceSerializeByMemcpy<T>`.

## Testing and Documentation

- Includes integration tests... but more, better tests need to be added
- Needs edits to existing public API documentation.

ShadauxCat committed:
WIP: fixing crashes when passing a NativeList to an RPC

ShadauxCat committed:
continued WIP, not ready for anything.

ShadauxCat committed:
Finally got it working correctly, still need to write tests.

ShadauxCat committed:
Added tests for templated types, fixed some issues around the ILPP generated for them.

ShadauxCat committed:
- Fixed failing tests (FixedString is not natively supported anymore without a ForceSerializeByMemcpy wrapper due to the ISerializeByMemcpy tag requirement)
- Made sure there's at least one test exercising Unity types going through RPCs... more tests exercising all of these are needed later.

ShadauxCat committed:
Merge branch 'develop' into fix/NativeList_crashes_when_used_in_RPCs

# Conflicts:
#	com.unity.netcode.gameobjects/Runtime/Messaging/Messages/SnapshotDataMessage.cs
#	com.unity.netcode.gameobjects/Runtime/NetworkVariable/NetworkVariable.cs
#	testproject/Packages/packages-lock.json

ShadauxCat committed:
Removed some debug logging.

ShadauxCat committed:
Standards

ShadauxCat committed:
Changelog

ShadauxCat committed:
Fix testproject-tools-integration

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/AssemblyInfo.cs:
```diff
@@ -5,6 +5,8 @@
 [assembly: InternalsVisibleTo("Unity.Netcode.Editor")]
 [assembly: InternalsVisibleTo("TestProject.EditorTests")]
+[assembly: InternalsVisibleTo("Unity.Netcode.Editor.CodeGen")]
+
```
```suggestion
[assembly: InternalsVisibleTo("Unity.Netcode.Editor.CodeGen")]
```

ShadauxCat replied:
I'm confused... these look identical.

0xFA11 replied:
deleted 1 extra empty line :P

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/ISerializeByMemcpy.cs:
```diff
@@ -0,0 +1,7 @@
+
+namespace Unity.Netcode
+{
+    public interface ISerializeByMemcpy
```
```suggestion
    public interface INetworkSerializeByMemcpy
```

ShadauxCat replied:
Are you at all worried this will make it easier to confuse this with `INetworkSerializable`?

0xFA11 replied:
I see more explicit intention by the name & consistency in `INetworkSerializeByMemcpy`

NoelStephensUnity replied:
On one hand, it does live here: `Unity.Netcode.ISerializeByMemcpy`
Which provides an understood intention.
On the other hand, `INetworkSerializeByMemcpy` does help further define that it is network serialization specific..
¯\\_ (ツ)_/¯

Rosme replied:
Where it lives doesn't really matter to me. At a glance, if I see `INetworkSerializeByMemcpy`, I know automatically it's for Networking. If I see `ISerializeByMemcpy` I might assume it's a generic thing for any and all scenario. And since we already have a `INetworkSerializable` for consistency and to fit, I would name it `INetworkSerializeByMemcpy`. I thing both are distinct enough to not be confused but understanding they are under the same umbrella.

ShadauxCat replied:
'twas renamed

0xFA11 commented on testproject/ProjectSettings/BurstAotSettings_StandaloneWindows.json:
```diff
@@ -1,6 +1,6 @@
 {
   "MonoBehaviour": {
-    "Version": 3,
+    "Version": 4,
```
prob no need changes in this file

ShadauxCat replied:
Ugh, I did it again...

0xFA11 commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@ Additional documentation and release notes are available at [Multiplayer Documen
 ### Added
 
 ### Changed
+- `unmanaged` structs are no longer universally accepted as RPC parameters because some structs (i.e., structs with pointers in them, such as `NativeList`) can't be supported by the universal struct serializer. Structs that are intended to be serialized across the network must add `ISerializeByMemcpy` to the interface list (i.e., `struct Foo : ISerializeByMemcpy`). This interface is empty and jus ...
```
doesn't this also mean that any existing code relying on a struct being serialized without the interface will immediately break after this change?

ShadauxCat replied:
Yep. But there's really no way to make `NativeList <T>` an error without making other things an error. We can choose to leave this part of it out, in which case we can fix template types causing crashes, but people will still be able to pass in NativeList and it will still crash... the alternative to doing this is documentation saying not to use NativeList in RPCs

ShadauxCat replied:
or we could add explicit support for NativeList and then that'll work, but NativeArray will crash... we can support all the things but there will always be something users can pass in without realizing it's invalid that will cause a crash.

0xFA11 commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@ Additional documentation and release notes are available at [Multiplayer Documen
 ### Added
 
 ### Changed
+- `unmanaged` structs are no longer universally accepted as RPC parameters because some structs (i.e., structs with pointers in them, such as `NativeList`) can't be supported by the universal struct serializer. Structs that are intended to be serialized across the network must add `ISerializeByMemcpy` to the interface list (i.e., `struct Foo : ISerializeByMemcpy`). This interface is empty and jus ...
```
```suggestion
- `unmanaged` structs are no longer universally accepted as RPC parameters because some structs (i.e., structs with pointers in them, such as `NativeList<T>`) can't be supported by the default memcpy struct serializer. Structs that are intended to be serialized across the network must add `INetworkSerializeByMemcpy` to the interface list (i.e., `struct Foo : INetworkSerializeByMemcpy`). This interface is empty and just serves to mark the struct as compatible with memcpy serialization. For external structs you can't edit, you can pass them to RPCs by wrapping them in `ForceNetworkSerializeByMemcpy<T>`. (#1901)
```

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/ForceSerializeByMemcpy.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace Unity.Netcode
+{
+    public struct ForceSerializeByMemcpy<T> : ISerializeByMemcpy, IEquatable<ForceSerializeByMemcpy<T>> where T : unmanaged, IEquatable<T>
```
```suggestion
    public struct ForceNetworkSerializeByMemcpy<T> : INetworkSerializeByMemcpy, IEquatable<ForceNetworkSerializeByMemcpy<T>> where T : unmanaged, IEquatable<T>
```

0xFA11 committed:
Merge branch 'develop' into fix/NativeList_crashes_when_used_in_RPCs

ShadauxCat committed:
Slight refactor to let ILPP find more types it needs to generate initializers for by looking for instances of NetworkVariable<T> and NetworkList<T>. This was required because runtime reflection doesn't work in AoT consoles; the runtime reflection will now only ever be needed if someone does something particularly odd like, for some reason, making a NetworkVariable or NetworkList that's not a field of a NetworkBehaviour.

Made an class to go between NetworkVariable and NetworkVariableBase so that future uses for these serialization functions will also be picked up automatically by ILPP.

Also, review feedback:
- Renamed ISerializeByMemcpy to INetworkSerializeByMemcpy
- Reverted Burst json file
- Removed an errant empty line

ShadauxCat committed:
Merge remote-tracking branch 'origin/fix/NativeList_crashes_when_used_in_RPCs' into fix/NativeList_crashes_when_used_in_RPCs

ShadauxCat committed:
Update com.unity.netcode.gameobjects/CHANGELOG.md

Co-authored-by: Fatih Mar <mfatihmar@gmail.com>

ShadauxCat committed:
Merge branch 'develop' into fix/NativeList_crashes_when_used_in_RPCs

NoelStephensUnity reviewed (approved):
LGTM!
Other than the standards check issue.  😸

ShadauxCat committed:
Standards.

ShadauxCat committed:
Merge remote-tracking branch 'origin/fix/NativeList_crashes_when_used_in_RPCs' into fix/NativeList_crashes_when_used_in_RPCs

0xFA11 committed:
Merge branch 'develop' into fix/NativeList_crashes_when_used_in_RPCs

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/ForceNetworkSerializeByMemcpy.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace Unity.Netcode
+{
+    public struct ForceNetworkSerializeByMemcpy<T> : INetworkSerializeByMemcpy, IEquatable<ForceNetworkSerializeByMemcpy<T>> where T : unmanaged, IEquatable<T>
```
we need xmldoc

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/INetworkSerializeByMemcpy.cs:
```diff
@@ -0,0 +1,7 @@
+
+namespace Unity.Netcode
+{
+    public interface INetworkSerializeByMemcpy
```
we need xmldoc

0xFA11 committed:
super minor touch

0xFA11 reviewed (changes requested):
we need xmldoc for new public `ForceNetworkSerializeByMemcpy` & `INetworkSerializeByMemcpy` types

ShadauxCat committed:
xmldocs and comments

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferWriter.cs:
```diff
@@ -777,6 +777,11 @@ public unsafe void CopyFrom(FastBufferWriter other)
             }
         }
 
+        // These structs enable overloading of WriteValue with different generic constraints.
+        // The compiler's actually able to distinguish between overloads based on generic constraints.
+        // But at the bytecode level, the constraints aren't included in the method signature.
+        // By adding a second parameter with a defaulted value, the signatures of each generic are different,
+        // thus allowing overloads of methods based on the first parameter meeting constraints.
```
should this be an xmldoc (aka 3 slashes summary block `///`)?

ShadauxCat replied:
Didn't intend it to be since it's referring to a whole section and not a specific struct or method.

I wasn't really sure what to do with xmldocs for this section because the methods are one line each and adding a 5 line xmldoc for each one seemed overkill.

0xFA11 reviewed (approved):
gtg 🚀

ShadauxCat committed:
Merge branch 'develop' into fix/NativeList_crashes_when_used_in_RPCs
